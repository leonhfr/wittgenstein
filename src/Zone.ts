// Packages
import * as turfHelpers from '@turf/helpers';
import * as turfInvariant from '@turf/invariant';

// Internal.
import * as Types from './types';

// Code.
export class Zone {
  readonly id: string;
  readonly area: string;
  readonly bbox: turfHelpers.BBox;
  readonly zone?: turfHelpers.Feature<
    turfHelpers.Polygon | turfHelpers.MultiPolygon
  >;

  static create(input: unknown): Zone | Error {
    if (isZone(input)) {
      return new Zone(input);
    }

    const errMsg = isSafeZone(input).errMsg;
    return new Error(`ValidationError: ${errMsg}`);
  }

  private constructor(input: CreateZoneInput) {
    this.id = input.id;
    this.area = input.area;
    this.bbox = input.bbox;
    this.zone = input.zone;
  }
}

// Validation.

export const ZONE_PROPS = ['id', 'area', 'bbox', 'zone'];

export const isZone = (input: unknown): input is CreateZoneInput => {
  return isSafeZone(input).isSafe;
};

export const isSafeZone = (input: unknown): Types.IsSafe => {
  if (typeof input !== 'object') {
    return {
      isSafe: false,
      errMsg: `Expected type of input to be object, got ${typeof input}.`,
    };
  }

  if (input === null) {
    return {
      isSafe: false,
      errMsg: `Expected input not to be null.`,
    };
  }

  for (const key of ZONE_PROPS) {
    if (!input.hasOwnProperty(key)) {
      return {
        isSafe: false,
        errMsg: `Expected all input properties to be set, missing ${key}.`,
      };
    }
  }

  const { id, area, bbox, zone } = input as {
    id: unknown;
    area: unknown;
    bbox: unknown;
    zone: unknown;
  };

  const strings = { id, area };

  // Strings

  for (const [key, value] of Object.entries(strings)) {
    if (typeof value !== 'string') {
      return {
        isSafe: false,
        errMsg: `Expected type of input.${key} to be string, got ${typeof value}.`,
      };
    }
  }

  // bbox

  if (!Array.isArray(bbox) || bbox.length !== 4) {
    return {
      isSafe: false,
      errMsg: `Expected type of input.bbox to be turf.helpers.BBox, got ${typeof bbox}.`,
    };
  }

  for (const item of bbox) {
    if (typeof item !== 'number') {
      return {
        isSafe: false,
        errMsg: `Expected type of input.bbox elements to be number, got ${typeof item} for one of them.`,
      };
    }
  }

  // zone?

  if (zone !== undefined) {
    try {
      turfInvariant.featureOf(
        zone as turfHelpers.Feature<any>,
        'Polygon',
        'isSafeZone'
      );
    } catch (err) {
      try {
        turfInvariant.featureOf(
          zone as turfHelpers.Feature<any>,
          'MultiPolygon',
          'isSafeZone'
        );
      } catch (err) {
        return {
          isSafe: false,
          errMsg: `Expected type of input.zone to be Feature<Polygon | MultiPolygon>. Turf.js threw ${
            err.name
          }: ${err.message}`,
        };
      }
    }
  }

  return {
    isSafe: true,
    errMsg: '',
  };
};

// Interface

export interface CreateZoneInput {
  readonly id: string;
  readonly area: string;
  readonly bbox: turfHelpers.BBox;
  readonly zone?: turfHelpers.Feature<
    turfHelpers.Polygon | turfHelpers.MultiPolygon
  >;
}
