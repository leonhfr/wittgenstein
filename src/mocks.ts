// Internal.
import { CreateAreaInput } from './Area';
import { CreatePhotoInput } from './Photo';

// Code.
export const createMockAreaInput = (area?: any): CreateAreaInput => {
  const _area = area || {};
  return {
    id: _area.id || '8ea8f412-4ee8-4f1d-9228-4c133b8fbd37',
    name: _area.name || 'Sant Marti',
    file: _area.file || 'sant-marti',
    lastScheduledAt: _area.lastScheduledAt || 0,
    refreshRate: _area.refreshRate || 86400,
    enabled: _area.enabled || true,
    zonesComputed: _area.zonesComputed || false,
  };
};

export const createMockPhotoInput = (photo?: any): CreatePhotoInput => {
  const _photo = photo || {};
  return {
    id: _photo.id || '32457360587',
    owner: _photo.owner || '41710899@N08',
    secret: _photo.secret || '69b27ceb18',
    server: _photo.server || '4898',
    farm: _photo.farm || 5,
    title: _photo.title || 'Arc de Triomf',
    description:
      _photo.description ||
      'The Arc de Triomf was built as the entrance of the 1888 Barcelona Universal Exposition. It was designed by architect Josep Vilaseca i Casanovas in the Neo-Mudéjar style.',
    ownername: _photo.ownername || 'jdf_92',
    views: _photo.views || 16,
    tags: _photo.tags || [
      'spain',
      'barcelona',
      'arcdetriomf',
      '1888barcelonaworldfair',
      'neomudéjar',
      '1888barcelonauniversalexposition',
      'catelonia',
    ],
    latitude: _photo.latitude || 41.390791,
    longitude: _photo.longitude || 2.18114,
    context: _photo.context || 0,
  };
};
