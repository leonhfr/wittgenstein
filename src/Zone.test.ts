// Internal.
import { Zone, ZONE_PROPS, isZone } from './Zone';
import * as Mocks from './mocks';

// Code.
describe('Zone', () => {
  describe('Class', () => {
    it('should be defined', () => {
      expect(Zone).toBeDefined();
    });

    describe('create', () => {
      it('should fail to create a new instance - invalid input', () => {
        expect(Zone.create(null)).toMatchSnapshot();
      });
      it('should create a new instance', () => {
        expect(Zone.create(Mocks.createMockZoneInput())).toMatchSnapshot();
      });
      it('should create a new instance - zone undefined', () => {
        expect(
          Zone.create({ ...Mocks.createMockZoneInput(), zone: undefined })
        ).toMatchSnapshot();
      });
      it('should create a new instance - MultiPolygon', () => {
        expect(
          Zone.create(Mocks.createMockZoneMultiPolygonInput())
        ).toMatchSnapshot();
      });
    });
  });

  describe('Validation', () => {
    it('should be defined', () => {
      expect(ZONE_PROPS).toBeDefined();
    });
    it('should match the expected value', () => {
      expect(ZONE_PROPS).toMatchSnapshot();
    });
  });

  describe('isZone', () => {
    it('should be defined', () => {
      expect(isZone).toBeDefined();
    });
    it('should handle an invalid input - wrong type', () => {
      expect(isZone('__FAIL__')).toBe(false);
    });
    it('should handle an invalid input - null type', () => {
      expect(isZone(null)).toBe(false);
    });
    it('should handle an invalid input - missing props', () => {
      expect(isZone({})).toBe(false);
    });
    it('should handle an invalid input - strings', () => {
      expect(
        isZone({
          ...Mocks.createMockZoneInput(),
          id: false,
        })
      ).toBe(false);
    });
    it('should handle an invalid input - arrays', () => {
      expect(
        isZone({
          ...Mocks.createMockZoneInput(),
          area: false,
        })
      ).toBe(false);
    });
    it('should handle an invalid input - bbox', () => {
      expect(
        isZone({
          ...Mocks.createMockZoneInput(),
          bbox: false,
        })
      ).toBe(false);
      expect(
        isZone({
          ...Mocks.createMockZoneInput(),
          bbox: [0, 0, 0],
        })
      ).toBe(false);
      expect(
        isZone({
          ...Mocks.createMockZoneInput(),
          bbox: [0, 0, 0, false],
        })
      ).toBe(false);
    });
    it('should handle an invalid input - zone', () => {
      expect(
        isZone({
          ...Mocks.createMockZoneInput(),
          zone: undefined,
        })
      ).toBe(true);
      expect(
        isZone({
          ...Mocks.createMockZoneInput(),
          zone: {},
        })
      ).toBe(false);
      expect(
        isZone({
          ...Mocks.createMockZoneInput(),
          zone: {
            type: 'Feature',
            properties: {},
            geometry: {
              type: 'LineString',
              coordinates: [[2.19, 41.43], [2.23, 41.44]],
            },
          },
        })
      ).toBe(false);
    });
  });
});
