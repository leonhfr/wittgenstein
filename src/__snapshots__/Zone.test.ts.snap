// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Zone Class create should create a new instance - MultiPolygon 1`] = `
Zone {
  "area": "8ea8f412-4ee8-4f1d-9228-4c133b8fbd37",
  "bbox": Array [
    2.18,
    41.4,
    2.19,
    41.5,
  ],
  "id": "4ab7068b-6c6c-46d2-8009-1d7d1ab35a3b",
  "zone": Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            Array [
              2.18,
              41.4,
            ],
            Array [
              2.19,
              41.5,
            ],
            Array [
              2.19,
              41.4,
            ],
            Array [
              2.18,
              41.4,
            ],
          ],
        ],
        Array [
          Array [
            Array [
              2.18,
              41.4,
            ],
            Array [
              2.18,
              41.5,
            ],
            Array [
              2.19,
              41.5,
            ],
            Array [
              2.18,
              41.4,
            ],
          ],
        ],
      ],
      "type": "MultiPolygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
}
`;

exports[`Zone Class create should create a new instance - zone undefined 1`] = `
Zone {
  "area": "8ea8f412-4ee8-4f1d-9228-4c133b8fbd37",
  "bbox": Array [
    2.18,
    41.4,
    2.19,
    41.5,
  ],
  "id": "4ab7068b-6c6c-46d2-8009-1d7d1ab35a3b",
  "zone": undefined,
}
`;

exports[`Zone Class create should create a new instance 1`] = `
Zone {
  "area": "8ea8f412-4ee8-4f1d-9228-4c133b8fbd37",
  "bbox": Array [
    2.18,
    41.4,
    2.19,
    41.5,
  ],
  "id": "4ab7068b-6c6c-46d2-8009-1d7d1ab35a3b",
  "zone": Object {
    "geometry": Object {
      "coordinates": Array [
        Array [
          Array [
            2.18,
            41.4,
          ],
          Array [
            2.19,
            41.5,
          ],
          Array [
            2.19,
            41.4,
          ],
          Array [
            2.18,
            41.4,
          ],
        ],
      ],
      "type": "Polygon",
    },
    "properties": Object {},
    "type": "Feature",
  },
}
`;

exports[`Zone Class create should fail to create a new instance - invalid input 1`] = `[Error: ValidationError: Expected input not to be null.]`;

exports[`Zone Validation should match the expected value 1`] = `
Array [
  "id",
  "area",
  "bbox",
]
`;
